PyPDF2/_cmap.py

```diff
@@ -57,72 +122,104 @@ def parse_encoding(ft: DictionaryObject, space_code: int) -> Tuple[List[str], in
                 x += 1
+    if isinstance(encoding, list):
+        encoding = dict(zip(range(256), encoding))
     return encoding, space_code
 
 
 def parse_to_unicode(ft: DictionaryObject, space_code: int) -> Tuple[Dict, int]:
-    map_dict: Dict[Any, Any] = {}
+    map_dict: Dict[
+        Any, Any
+    ] = (
+        {}
+    )  # will store all translation code and map_dict[-1] we will have the number of bytes to convert
     if "/ToUnicode" not in ft:
-        return map_dict, space_code
+        return {}, space_code
     process_rg: bool = False
     process_char: bool = False
-    cm: str = cast(DecodedStreamObject, ft["/ToUnicode"]).get_data().decode("utf-8")
-    for l in (
+    cm: bytes = cast(DecodedStreamObject, ft["/ToUnicode"]).get_data()
+    # we need to prepare cm before due to missing return line in pdf printed to pdf from word
+    cm = (
         cm.strip()
-        .replace("<", " ")
-        .replace(">", "")
-        .replace("[", " [ ")
-        .replace("]", " ] ")
-        .split("\n")
-    ):
-        if l == "":
+        .replace(b"beginbfchar", b"\nbeginbfchar\n")
+        .replace(b"endbfchar", b"\nendbfchar\n")
+        .replace(b"beginbfrange", b"\nbeginbfrange\n")
+        .replace(b"endbfrange", b"\nendbfrange\n")
+        .replace(b"<<", b"\n{\n")  # text between << and >> not used but
+        .replace(b">>", b"\n}\n")  # some solution to find it back
+    )
+    ll = cm.split(b"<")
+    for i in range(len(ll)):
+        j = ll[i].find(b">")
+        if j >= 0:
+            ll[i] = ll[i][:j].replace(b" ", b"") + b" " + ll[i][j + 1 :]
+    cm = (b" ".join(ll)).replace(b"[", b" [ ").replace(b"]", b" ]\n ")
+
+    for l in cm.split(b"\n"):
+        if l == b"" or l == b" ":
             continue
-        if "beginbfrange" in l:
+        if b"beginbfrange" in l:
             process_rg = True
-        elif "endbfrange" in l:
+        elif b"endbfrange" in l:
             process_rg = False
-        elif "beginbfchar" in l:
+        elif b"beginbfchar" in l:
             process_char = True
-        elif "endbfchar" in l:
+        elif b"endbfchar" in l:
             process_char = False
         elif process_rg:
-            lst = [x for x in l.split(" ") if x]
+            lst = [x for x in l.split(b" ") if x]
             a = int(lst[0], 16)
             b = int(lst[1], 16)
-            if lst[2] == "[":
+            nbi = len(lst[0])
+            map_dict[-1] = nbi // 2
+            fmt = b"%%0%dX" % nbi
+            if lst[2] == b"[":
                 for sq in lst[3:]:
-                    if "]":
+                    if sq == b"]":
                         break
-                    map_dict[a] = unhexlify(sq).decode("utf-16-be")
+                    map_dict[
+                        unhexlify(fmt % a).decode(
+                            "charmap" if map_dict[-1] == 1 else "utf-16-be"
+                        )
+                    ] = unhexlify(sq).decode("utf-16-be")
```

Comment:
```suggestion
                    key = unhexlify(fmt % a).decode(
                        "charmap" if map_dict[-1] == 1 else "utf-16-be"
                    )
                    try:
                        map_dict[key] = unhexlify(fmt2 % c).decode("utf-16-be")
                    except UnicodeDecodeError:
                        map_dict[key] = key
```

Triggered by https://corpora.tika.apache.org/base/docs/govdocs1/971/971703.pdf

Reply:
Again, I'm pretty certain there is a better way to deal with that. I just don't know how.

Reply:
I've retested the file and (with latests mods) the issue is somewhere else 
```
  File "C:\Python310\lib\site-packages\PyPDF2\_page.py", line 1302, in extract_text
    return self._extract_text(self, self.pdf, space_width, PG.CONTENTS)
  File "C:\Python310\lib\site-packages\PyPDF2\_page.py", line 1284, in _extract_text
    process_operation(operator, operands)
  File "C:\Python310\lib\site-packages\PyPDF2\_page.py", line 1222, in process_operation
    t = tt.decode(cmap[0])  # apply str encoding
  File "C:\Python310\lib\encodings\utf_16_be.py", line 16, in decode
    return codecs.utf_16_be_decode(input, errors, True)
UnicodeDecodeError: 'utf-16-be' codec can't decode byte 0x20 in position 0: truncated data
```

For this latest issue, I've checked the PDF and the font transcoding do expect 2 bytes standard. The file at one place is showing only b" " (just before I've noted a <0055>). Don't know how to cope with this situation